runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java

```diff
@@ -864,7 +864,9 @@ private void pruneRemovedRouters(@Nonnull Layout layout) {
                         .contains(NodeLocator.getLegacyEndpoint(endpoint)))
                 .forEach(endpoint -> {
                     try {
-                        nodeRouterPool.getNodeRouters().get(endpoint).stop();
+                        // Simply remove the router from the pool and let finalize() method close
```

Comment:
but stop() method does some actions, do we lose those actions with this change?

Reply:
That's why I use finalize() to clean up. stop() is removed here because it can create a race where others are holding this router, and if it sends a message with this stoped router, it gets a ShutdownException.